openhands/memory/condenser.py

```diff
@@ -1,24 +1,412 @@
+        match config:
+            case NoOpCondenserConfig():
+                return NoOpCondenser()
+
+            case ObservationMaskingCondenserConfig():
+                return ObservationMaskingCondenser(
+                    **config.model_dump(exclude=['type'])
+                )
+
+            case RecentEventsCondenserConfig():
+                return RecentEventsCondenser(**config.model_dump(exclude=['type']))
+
+            case LLMSummarizingCondenserConfig(llm_config=llm_config):
+                return LLMSummarizingCondenser(llm=LLM(config=llm_config))
+
+            case AmortizedForgettingCondenserConfig():
+                return AmortizedForgettingCondenser(
+                    **config.model_dump(exclude=['type'])
+                )
+
+            case LLMAttentionCondenserConfig(llm_config=llm_config):
+                return LLMAttentionCondenser(
+                    llm=LLM(config=llm_config),
+                    **config.model_dump(exclude=['type', 'llm_config']),
+                )
+
+            case _:
+                raise ValueError(f'Unknown condenser config: {config}')
+
+
+class RollingCondenser(Condenser, ABC):
+    """Base class for a specialized condenser strategy that applies condensation to a rolling history.
+
+    The rolling history is computed by appending new events to the most recent condensation. For example, the sequence of calls::
+
+        assert state.history == [event1, event2, event3]
+        condensation = condenser.condensed_history(state)
+
+        # ...new events are added to the state...
+
+        assert state.history == [event1, event2, event3, event4, event5]
+        condenser.condensed_history(state)
+
+    will result in second call to `condensed_history` passing `condensation + [event4, event5]` to the `condense` method.
+    """
+
+    def __init__(self) -> None:
+        self._condensation: list[Event] = []
+        self._last_history_length: int = 0
+
+        super().__init__()
+
+    @override
+    def condensed_history(self, state: State) -> list[Event]:
+        new_events = state.history[self._last_history_length :]
+
+        with self.metadata_batch(state):
+            results = self.condense(self._condensation + new_events)
+
+        self._condensation = results
+        self._last_history_length = len(state.history)
+
+        return results
+
+
+@dataclass
+class CondensationObservation(Observation):
+    """The output of a condensation action."""
+
+    observation: str = ObservationType.RUN
+
+    @property
+    def message(self) -> str:
+        return self.content
+
+
+class NoOpCondenser(Condenser):
+    """A condenser that does nothing to the event sequence."""
+
+    def condense(self, events: list[Event]) -> list[Event]:
+        """Returns the list of events unchanged."""
+        return events
+
+
+class ObservationMaskingCondenser(Condenser):
+    """A condenser that masks the values of observations outside of a recent attention window."""
```

Comment:
Oh, that's very interesting! Surely it can help to remove from sight some of the 10k outdated tokens in an obs, while keeping the flow complete! This strategy is a new one, how were the tests on this?

Reply:
The idea suggested by @rbren. The results just came in this morning, I haven't run them through my evaluation script but a cursory glance shows an attention window of ~15 events gives identical performance to the baseline (no condensation) and uses less tokens overall.